aspnet/fundamentals/static-files.rst

```diff
@@ -1,8 +1,336 @@
+
+Using the UseFileServer method
+------------------------------
+
+In addition to the ``UseStaticFiles``, ``UseDefaultFiles``, and ``UseDirectoryBrowser`` extensions methods, there is also a single method - ``UseFileServer`` - that combines the functionality of all three methods. The following example code shows some common ways to use this method:
+
+.. code-block:: c#
+
+  // Enable all static file middleware (serving of static files and default files) EXCEPT directory browsing.
+  app.UseFileServer();
+
+.. code-block:: c#
+
+  // Enables all static file middleware (serving of static files, default files, and directory browsing).
+  app.UseFileServer(enableDirectoryBrowsing: true);
+
+As with the ``UseStaticFiles``, ``UseDefaultFiles``, and ``UseDirectoryBrowser`` methods, if you wish to serve files that exist outside the webroot, you instantiate and configure an "options" object that you pass as a parameter to ``UseFileServer``. For example, let's say you have the following directory hierarchy in your Web app:
+
+- wwwroot
+
+  - css
+  - images
+  - ...
+
+- MyStaticFiles
+
+  - test.png
+  - default.html
+
+Using the hierarchy example above, you might want to enable static files, default files, and browsing for the ``MyStaticFiles`` directory. In the following code snippet, that is accomplished with a single call to ``UseFileServer``.
+
+.. code-block:: c#
+
+  // Enable all static file middleware (serving of static files, default files,
+  // and directory browsing) for the MyStaticFiles directory.
+  app.UseFileServer(new FileServerOptions()
+  {
+      FileProvider = new PhysicalFileProvider(@"D:\Source\WebApplication1\src\WebApplication1\MyStaticFiles"),
+      RequestPath = new PathString("/StaticFiles"),
+      EnableDirectoryBrowsing = true
+  });
+
+Using the example hierarchy and code snippet from above, here's what happens if the user browses to various URIs:
+
+  - ``http://<yourApp>/StaticFiles/test.png`` - The ``MyStaticFiles/test.png`` file will be served to and presented by the browser.
+  - ``http://<yourApp>/StaticFiles`` - Since a default file is present (``MyStaticFiles/default.html``), that file will be served. If that file didn't exist, the browser would present a list of files in the ``MyStaticFiles`` directory (because the ``FileServerOptions.EnableDirectoryBrowsing`` property is set to ``true``).
+
+Working with content types
+--------------------------
+
+The ASP.NET static files middleware defines `almost 400 known file content types <https://github.com/aspnet/StaticFiles/blob/dev/src/Microsoft.AspNet.StaticFiles/FileExtensionContentTypeProvider.cs>`_. If the user attempts to reach a file of an unknown file type, ASP.NET will return a 404 Not Found error.
```

Comment:
The static file middleware will not attempt to serve the file and will pass control to the next middleware. This may eventually result in a 404 if no one handles the request.